app/scripts/migrations/151.ts

```diff
@@ -0,0 +1,310 @@
+      ),
+    );
+
+    return state;
+  }
+
+  if (
+    !selectedNetworkControllerState ||
+    !isObject(selectedNetworkControllerState)
+  ) {
+    global.sentry?.captureException?.(
+      new Error(
+        'Skipping migration: `SelectedNetworkController` state not found or is not an object.',
+      ),
+    );
+
+    return state;
+  }
+
+  if (!isObject(selectedNetworkControllerState.domains)) {
+    global.sentry?.captureException?.(
+      new Error(
+        'Skipping migration: `SelectedNetworkController.domains` state is not an object.',
+      ),
+    );
+
+    return state;
+  }
+
+  const currentChainId = getChainId(
+    networkControllerState.selectedNetworkClientId,
+    networkControllerState.networkConfigurationsByChainId,
+  );
+
+  if (typeof currentChainId !== 'string') {
+    global.sentry?.captureException?.(
+      new Error(
+        'Skipping migration: No currently selected chainId was found, or the chainId resolved from the `NetworkController.selectedNetworkClientId` value is not a string.',
+      ),
+    );
+
+    return state;
+  }
+
+  const updatedSubjects: string[] = [];
+
+  // Add permission to use the current globally selected network to all Snaps
+  // that have the `endowment:ethereum-provider` permission.
+  const entries = Object.entries(permissionControllerState.subjects) as [
+    string,
+    GenericPermissionControllerSubject,
+  ][];
+
+  permissionControllerState.subjects = entries.reduce<
+    PermissionControllerSubjects<PermissionConstraint>
+  >((accumulator, [key, subject]) => {
+    const permissionKeys = Object.keys(subject.permissions);
+    const needsMigration = permissionKeys.includes(
+      SnapEndowments.EthereumProvider,
+    );
+
+    if (!needsMigration) {
+      return {
+        ...accumulator,
+        [key]: subject,
+      };
+    }
+
+    updatedSubjects.push(key);
+
+    const caip25PermissionCaveatWithCurrentChainIdSet = addPermittedEthChainId(
+      getExistingCaip25PermissionCaveat(subject),
+      currentChainId,
+    );
+
+    const {
+      date = Date.now(),
+      id = nanoid(),
+      invoker = key,
+      parentCapability = Caip25EndowmentPermissionName,
+    } = subject.permissions[Caip25EndowmentPermissionName] ?? {};
+
+    const newSubject: GenericPermissionControllerSubject = {
+      ...subject,
+      permissions: {
+        ...subject.permissions,
+        [Caip25EndowmentPermissionName]: {
+          caveats: [
+            {
+              type: Caip25CaveatType,
+              value: caip25PermissionCaveatWithCurrentChainIdSet,
+            },
+          ],
+          date,
+          id,
+          invoker,
+          parentCapability,
+        },
+      },
+    };
+
+    return {
+      ...accumulator,
+      [key]: newSubject,
+    };
+  }, {});
+
+  const currentDomains = selectedNetworkControllerState.domains;
+  const domains = Object.fromEntries(
+    updatedSubjects.map((subject) => [subject, currentChainId]),
```

Comment:
pretty sure the value of this should be a networkClientId, not chainId

Reply:
oooof good catch Jiexi! yeah I think this should be something like:
```suggestion
    updatedSubjects.map((subject) => [subject, networkControllerState.selectedNetworkClientId]),
```

Reply:
I thought I checked that both work, but maybe not. Will update! The network controller and selected network controller can be hard to follow, especially in migrations where we can't really rely on the controllers directly 😅